emote/memory/coverage_based_strategy.py

```diff
@@ -0,0 +1,110 @@
+"""
+
+"""
+
+import random
+
+from typing import Sequence
+
+import numpy as np
+
+from .core_types import SamplePoint
+from .strategy import EjectionStrategy, SampleStrategy, Strategy
+
+
+class CoverageBasedStrategy(Strategy):
+    """A sampler intended to sample based on coverage of experiences,
+    favoring less-visited states. This base class can be used for implementing
+    various coverage-based sampling strategies."""
+
+    def __init__(self, alpha=0.5):
+        super().__init__()
+        self._identities = {}
+        self._sample_count = {}
+        self._ids = []
+        self._prios = []
+        self._dirty = False
+        self._alpha = alpha
+
+    def track(self, identity: int, sequence_length: int):
+        self._dirty = True
+        self._identities[identity] = sequence_length
+        self._sample_count[identity] = self._sample_count.get(identity, 0)
+
+    def forget(self, identity: int):
+        self._dirty = True
+        del self._identities[identity]
+        del self._sample_count[identity]
+
+    def _rebalance(self):
+        self._dirty = False
+        original_prios = np.array(tuple(self._identities.values())) / sum(self._identities.values())
+        self._ids = np.array(tuple(self._identities.keys()), dtype=np.int64)
+
+        sample_prios = np.array(
+            [1 / (self._sample_count[id] + 1) ** self._alpha for id in self._ids]
+        )
+        combined_prios = original_prios * sample_prios
+
+        sum_prios = sum(combined_prios)
+        self._prios = combined_prios / sum_prios
```

Comment:
I *think* this needs to account for length of episodes somehow, something like `sample_count[i] / len[i]` as a weight? So we get uniformity over transitions, and not just episodes.

Reply:
Isn't that being done through `combined_prios`? The multiplication should take care of that 🤔

Reply:
Right, yes.